helix-term/src/commands.rs

```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
 }
 
 // I inserts at the first nonwhitespace character of each line with a selection
+// If the line is empty, automatically indent
 fn insert_at_line_start(cx: &mut Context) {
-    goto_first_nonwhitespace(cx);
     enter_insert_mode(cx);
+
+    let (view, doc) = current!(cx.editor);
+
+    let text = doc.text().slice(..);
+    let contents = doc.text();
+    let selection = doc.selection(view.id);
+
+    let language_config = doc.language_config();
+    let syntax = doc.syntax();
+    let tab_width = doc.tab_width();
+
+    let mut ranges = SmallVec::with_capacity(selection.len());
+    let mut offs = 0;
+
+    let mut transaction = Transaction::change_by_selection(contents, selection, |range| {
+        let cursor = range.from();
+        let cursor_line = text.char_to_line(cursor);
+
+        let doc_end = contents.len_chars();
+        if range.from() == doc_end {
+            ranges.push(Range::point(doc_end + offs));
+            return (doc_end, doc_end, None);
+        }
+
+        let line_content = line_without_line_ending(&text, cursor_line);
+        if line_content.len_chars() == 0 {
+            // line is empty => auto indent
+            let (line_end_index, line_end_offset_width) = if cursor_line == 0 {
+                (0, 0)
+            } else {
+                (
+                    line_end_char_index(&doc.text().slice(..), cursor_line),
+                    doc.line_ending.len_chars(),
+                )
+            };
+
+            let indent = indent::indent_for_newline(
+                language_config,
+                syntax,
+                &doc.indent_style,
+                tab_width,
+                text,
+                cursor_line,
+                line_end_index,
+                cursor_line,
+            );
+
+            let indent_len = indent.len();
+            let mut text = String::with_capacity(indent_len);
+            text.push_str(&indent);
+
+            // calculate new selection ranges
+            let pos = offs + line_end_index + line_end_offset_width;
+            ranges.push(Range::point((pos + indent_len).saturating_sub(1)));
+
+            offs += text.chars().count();
```

Comment:
```suggestion
            let pos = offs + cursor_line_start ;
            let indent_width = indent.chars().count();
            ranges.push(pos + indent_width);
            offs += indent_width;
```

you are using byte indicies here (`indent_len`). You should be using char incidences as you do below. Also you are not considering select mode here.

Reply:
Done.